### PR TITLE
[api] add a long poll API /transactions/wait_by_hash/:txn_hash

### PR DESCRIPTION
**Pull request by bchocho**
### Description

The motivation is to reduce e2e latencies by learning of committed responses immediately. Currently, it is up to the client to poll /transactions/by_hash and there is overhead on the poll intervals and the RTT of the poll.

There is concern around holding large numbers of connections for an extended time -- to alleviate this we put a limit on active long polls, and simply return immediately when that limit is reached. The API does not give guarantees about how long the long poll will wait.

### Test Plan

Added unit tests. Deployed locally and in previewnet and tested long poll behavior.

### CODE REVIEW

File: api/src/transactions.rs [hunk 1]
```diff
@@ -209,6 +209,44 @@ impl TransactionsApi {
...
+        let start_time = std::time::Instant::now();
+        loop {
+            return match self
+                .get_transaction_by_hash_inner(&accept_type, txn_hash.0)
+                .await
+            {
+                Ok(txn) => Ok(txn),
+                Err(err) => {
+                    if let BasicErrorWith404::NotFound(..) = &err {
+                        // For testing, wait 10 minutes
```
Comment (sitalkedia) [inline api/src/transactions.rs:238]: can you explain why is this needed?

Comment (bchocho) [reply api/src/transactions.rs:250]: Ah, this was just for some local testing. Will set at 2 seconds

**Review by gregnazario — COMMENTED** [review]
Have we considered web sockets rather than blocking API requests?

cc @gedigi

**Comment by bchocho** [comment]
I feel like web sockets is overkill for this specific scenario, where on average we should be returning a response within 2 seconds.

If there are more use cases for async events, we should consider it. Using it will require extensive changes on both the API side and SDK side as well.

**Review by manudhundi — APPROVED** [review]

File: api/src/transactions.rs [hunk 2]
```diff
@@ -209,6 +210,59 @@ impl TransactionsApi {
...
+    /// The request will return immediately once the committed transaction is found.
+    /// If the committed transaction is not found within the timeout, a 404 will be returned.
+    /// In case of other errors, the request will return immediately with the error.
+    #[oai(
+        path = "/transactions/wait_by_hash/:txn_hash",
+        method = "get",
+        operation_id = "wait_transaction_by_hash",
+        tag = "ApiTags::Transactions"
+    )]
+    async fn wait_transaction_by_hash(
```
Comment (sitalkedia) [inline api/src/transactions.rs:226]: Do you know if there is any limit on number of concurrent connection the server can support? A concern might be that we exhaust that limit if we have too many of this blocking call. Also, a similar issue might happen for the threadpool exhaustion - can we make this call in the context of a blocking threadpool?

File: api/src/transactions.rs [hunk 3]
```diff
@@ -209,6 +210,73 @@ impl TransactionsApi {
...
+                .observe(0.0);
+            return self
+                .get_transaction_by_hash_inner(&accept_type, txn_hash.0)
+                .await;
+        }
+
+        let start_time = std::time::Instant::now();
+        self.context
+            .wait_for_hash_active_connections
+            .fetch_add(1, std::sync::atomic::Ordering::Relaxed);
```
Comment (sitalkedia) [inline api/src/transactions.rs:258]: Hmm there is a race condition here as we increment after checking the counter. I think it is okay as we don't have to guarantee that the active connections is strictly less than the threshold but let's add a comment here stating that is intentional.

**Review by sitalkedia — APPROVED** [review]
LGTM % minor nit. We will need to follow up to make changes on SDK to use the newly added API.

**Comment by bchocho** [comment]
@gedigi 
* This is async code, so we won't block threads
* We have a limit of the number of long poll calls now, so it addresses the concern around the number of connections this can tie up
